src/mongo/shell.ts

```diff
@@ -72,43 +91,64 @@ export class Shell {
 	async exec(script: string): Promise<string> {
 		script = this.convertToSingleLine(script);
 		const executionId = this._generateExecutionSequenceId();
+		// tslint:disable-next-line:no-any
+		let writeError: any;
 
 		try {
 			this.mongoShell.stdin.write(script, 'utf8');
 			this.mongoShell.stdin.write(os.EOL);
+
+			// Write out a unique number into stdin as a sentinel. We will know we've seen the end of the
+			//   result data when we get this number back out at the end of the data.
 			this.mongoShell.stdin.write(executionId, 'utf8');
 			this.mongoShell.stdin.write(os.EOL);
 		} catch (error) {
-			window.showErrorMessage(error.toString());
+			// Generally if writing to the process' stdin fails it has already exited
+			// with an error, and we will get notification via its stdout. So hold on to
+			// this and only return it if we don't see any other notifications after timeout.
+			writeError = error;
 		}
 
-		return await new Promise<string>((c, e) => {
+		return await new Promise<string>((resolve, reject) => {
 			let executed = false;
-			const handler = setTimeout(
+			const timeoutHandler = setTimeout(
 				() => {
 					if (!executed) {
-						e(`Timed out executing MongoDB command "${script}"`);
+						if (writeError) {
+							reject(writeError);
+						} else {
+							reject(`Timed out executing MongoDB command "${script}"`);
+						}
 					}
 				},
 				5000);
-			const disposable = this.onResult.event(result => {
+
+			// Handle results
+			const disposable = this.onResult.event((result: CommandResult) => {
+				// We will only process the first time onResult is fired
 				disposable.dispose();
 
-				if (result && result.code) {
+				if (result.stringResult) {
+					let lines = result.stringResult.split(os.EOL).filter(line => !!line && line !== 'Type "it" for more');
+					lines = lines[lines.length - 1] === 'Type "it" for more' ? lines.splice(lines.length - 1, 1) : lines;
```

Comment:
Given that you filter out `Type "it" for more` one line above, in what scenario would the true branch execute here?